File: application/src/test/java/de/seuhd/campuscoffee/tests/acceptance/CucumberReviewSteps.java
package de.seuhd.campuscoffee.tests.acceptance;

import de.seuhd.campuscoffee.api.dtos.PosDto;
import de.seuhd.campuscoffee.api.dtos.ReviewDto;
import de.seuhd.campuscoffee.api.dtos.UserDto;
import de.seuhd.campuscoffee.api.mapper.PosDtoMapper;
import de.seuhd.campuscoffee.domain.tests.TestFixtures;
import io.cucumber.java.DataTableType;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static de.seuhd.campuscoffee.tests.SystemTestUtils.Requests.posRequests;
import static de.seuhd.campuscoffee.tests.SystemTestUtils.Requests.reviewRequests;
import static de.seuhd.campuscoffee.tests.SystemTestUtils.Requests.userRequests;
import static org.assertj.core.api.Assertions.assertThat;

/**
 * Step definitions for the review approval Cucumber tests. The Spring context, container, and
 * cleanup hooks live in {@link CucumberSpringConfiguration}.
 */
public class CucumberReviewSteps {

    @Autowired
    protected PosDtoMapper posDtoMapper;

    private final Map<String, UserDto> usersByLogin = new HashMap<>();
    private final Map<String, PosDto> posByName = new HashMap<>();
    private final Map<String, ReviewDto> reviewsByAuthorAndPos = new HashMap<>();
    private Integer lastApprovalStatusCode;

    /**
     * Register a Cucumber DataTable type for UserDto (distinct return type from
     * {@link CucumberPosSteps#toPosDto}, so there is no registration clash).
     *
     * @param row the DataTable row to map to a UserDto object
     * @return the mapped UserDto object
     */
    @DataTableType
    @SuppressWarnings("unused")
    public UserDto toUserDto(Map<String, String> row) {
        return UserDto.builder()
                .loginName(row.get("loginName"))
                .emailAddress(row.get("emailAddress"))
                .firstName(row.get("firstName"))
                .lastName(row.get("lastName"))
                .build();
    }

    // Given -----------------------------------------------------------------------

    @Given("the following users exist:")
    public void theFollowingUsersExist(List<UserDto> users) {
        userRequests.create(users).forEach(user -> usersByLogin.put(user.loginName(), user));
    }

    @Given("a POS named {string} exists")
    public void aPosNamedExists(String name) {
        PosDto pos = posDtoMapper.fromDomain(TestFixtures.getPosFixturesForInsertion().getFirst())
                .toBuilder().name(name).build();
        posByName.put(name, posRequests.create(List.of(pos)).getFirst());
    }

    @Given("{string} reviewed {string} with {string}")
    public void userReviewedPosWith(String login, String posName, String text) {
        createReview(login, posName, text);
    }

    // When -----------------------------------------------------------------------

    @When("{string} reviews {string} with {string}")
    public void userReviewsPosWith(String login, String posName, String text) {
        createReview(login, posName, text);
    }

    @When("{string} approves the review by {string} for {string}")
    public void userApprovesReview(String approverLogin, String authorLogin, String posName) {
        ReviewDto review = reviewsByAuthorAndPos.get(reviewKey(authorLogin, posName));
        ReviewDto updated = reviewRequests.approve(review.getId(), usersByLogin.get(approverLogin).getId());
        reviewsByAuthorAndPos.put(reviewKey(authorLogin, posName), updated);
    }

    @When("{string} tries to approve the review by {string} for {string}")
    public void userTriesToApproveReview(String approverLogin, String authorLogin, String posName) {
        ReviewDto review = reviewsByAuthorAndPos.get(reviewKey(authorLogin, posName));
        lastApprovalStatusCode = reviewRequests
                .approveAndReturnStatusCode(review.getId(), usersByLogin.get(approverLogin).getId());
    }

    // Then -----------------------------------------------------------------------

    @Then("the review by {string} for {string} is approved")
    public void theReviewIsApproved(String authorLogin, String posName) {
        assertThat(currentReview(authorLogin, posName).approved()).isTrue();
    }

    @Then("the review by {string} for {string} is not approved")
    public void theReviewIsNotApproved(String authorLogin, String posName) {
        assertThat(currentReview(authorLogin, posName).approved()).isFalse();
    }

    @Then("the approval is rejected")
    public void theApprovalIsRejected() {
        assertThat(lastApprovalStatusCode).isEqualTo(HttpStatus.BAD_REQUEST.value());
    }

    // helpers ---------------------------------------------------------------------

    private void createReview(String login, String posName, String text) {
        ReviewDto review = ReviewDto.builder()
                .posId(posByName.get(posName).getId())
                .authorId(usersByLogin.get(login).getId())
                .review(text)
                .build();
        reviewsByAuthorAndPos.put(reviewKey(login, posName), reviewRequests.create(List.of(review)).getFirst());
    }

    private ReviewDto currentReview(String authorLogin, String posName) {
        return reviewRequests.retrieveById(reviewsByAuthorAndPos.get(reviewKey(authorLogin, posName)).getId());
    }

    private String reviewKey(String login, String posName) {
        return login + " @ " + posName;
    }
}

File: application/src/test/kotlin/de/seuhd/campuscoffee/tests/acceptance/CucumberReviewSteps.kt
package de.seuhd.campuscoffee.tests.acceptance

import de.seuhd.campuscoffee.api.dtos.PosDto
import de.seuhd.campuscoffee.api.dtos.ReviewDto
import de.seuhd.campuscoffee.api.dtos.UserDto
import de.seuhd.campuscoffee.api.mapper.PosDtoMapper
import de.seuhd.campuscoffee.domain.tests.TestFixtures
import de.seuhd.campuscoffee.tests.SystemTestUtils.Credentials
import de.seuhd.campuscoffee.tests.SystemTestUtils.MODERATOR
import de.seuhd.campuscoffee.tests.SystemTestUtils.posRequests
import de.seuhd.campuscoffee.tests.SystemTestUtils.reviewRequests
import de.seuhd.campuscoffee.tests.SystemTestUtils.userRequests
import io.cucumber.java.DataTableType
import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.assertj.core.api.Assertions.assertThat
import org.springframework.http.HttpStatus

/**
 * Step definitions for the review approval Cucumber tests. The Spring context, container, and cleanup
 * hooks live in [CucumberSpringConfiguration].
 *
 * Every write request authenticates. The feature's users are registered with a known password (the Background
 * table carries it), and the steps remember each login's credentials so a review is created as its named
 * author and each approval is recorded under the approver's own login.
 */
class CucumberReviewSteps(
    private val posDtoMapper: PosDtoMapper
) {
    private val credentialsByLogin = mutableMapOf<String, Credentials>()
    private val posByName = mutableMapOf<String, PosDto>()
    private val reviewsByAuthorAndPos = mutableMapOf<String, ReviewDto>()
    private var lastApprovalStatusCode = 0

    /**
     * Register a Cucumber DataTable type for [UserDto] (distinct return type from
     * [CucumberPosSteps.toPosDto], so there is no registration clash). The `password` column carries the
     * known password the steps later authenticate with.
     *
     * @param row the DataTable row to map to a [UserDto]
     * @return the mapped [UserDto]
     */
    @DataTableType
    @Suppress("unused")
    fun toUserDto(row: Map<String, String>): UserDto =
        UserDto(
            loginName = row["loginName"],
            emailAddress = row["emailAddress"],
            firstName = row["firstName"],
            lastName = row["lastName"],
            password = row["password"]
        )

    // Given -----------------------------------------------------------------------

    @Given("the following users exist:")
    fun theFollowingUsersExist(users: List<UserDto>) {
        userRequests.create(users).forEachIndexed { index, created ->
            // the response never carries the password, so pair the login with the one from the table row
            credentialsByLogin[created.loginName!!] = Credentials(created.loginName!!, users[index].password!!)
        }
    }

    @Given("a POS named {string} exists")
    fun aPosNamedExists(name: String) {
        val pos = posDtoMapper.fromDomain(TestFixtures.getPosFixturesForInsertion().first()).copy(name = name)
        // creating a POS requires a moderator
        posByName[name] = posRequests.create(listOf(pos), MODERATOR).first()
    }

    @Given("{string} reviewed {string} with {string}")
    fun userReviewedPosWith(
        login: String,
        posName: String,
        text: String
    ) {
        createReview(login, posName, text)
    }

    // When -----------------------------------------------------------------------

    @When("{string} reviews {string} with {string}")
    fun userReviewsPosWith(
        login: String,
        posName: String,
        text: String
    ) {
        createReview(login, posName, text)
    }

    @When("{string} approves the review by {string} for {string}")
    fun userApprovesReview(
        approverLogin: String,
        authorLogin: String,
        posName: String
    ) {
        val review = reviewsByAuthorAndPos.getValue(reviewKey(authorLogin, posName))
        val updated = reviewRequests.approve(review.id!!, credentialsByLogin.getValue(approverLogin))
        reviewsByAuthorAndPos[reviewKey(authorLogin, posName)] = updated
    }

    @When("{string} tries to approve the review by {string} for {string}")
    fun userTriesToApproveReview(
        approverLogin: String,
        authorLogin: String,
        posName: String
    ) {
        val review = reviewsByAuthorAndPos.getValue(reviewKey(authorLogin, posName))
        lastApprovalStatusCode =
            reviewRequests.approveAndReturnStatusCode(review.id!!, credentialsByLogin.getValue(approverLogin))
    }

    // Then -----------------------------------------------------------------------

    @Then("the review by {string} for {string} is approved")
    fun theReviewIsApproved(
        authorLogin: String,
        posName: String
    ) {
        assertThat(currentReview(authorLogin, posName).approved).isTrue()
    }

    @Then("the review by {string} for {string} is not approved")
    fun theReviewIsNotApproved(
        authorLogin: String,
        posName: String
    ) {
        assertThat(currentReview(authorLogin, posName).approved).isFalse()
    }

    @Then("the approval is rejected")
    fun theApprovalIsRejected() {
        assertThat(lastApprovalStatusCode).isEqualTo(HttpStatus.BAD_REQUEST.value())
    }

    // helpers ---------------------------------------------------------------------

    private fun createReview(
        login: String,
        posName: String,
        text: String
    ) {
        // the author is the authenticated caller, so the body carries no authorId; the review is created
        // as the named login
        val review = ReviewDto(posId = posByName.getValue(posName).id, review = text)
        reviewsByAuthorAndPos[reviewKey(login, posName)] =
            reviewRequests.create(listOf(review), credentialsByLogin.getValue(login)).first()
    }

    private fun currentReview(
        authorLogin: String,
        posName: String
    ): ReviewDto = reviewRequests.retrieveById(reviewsByAuthorAndPos.getValue(reviewKey(authorLogin, posName)).id!!)

    private fun reviewKey(
        login: String,
        posName: String
    ): String = "$login @ $posName"
}
